Add support for Strimzi Metrics Reporter (#954)
strimzi/strimzi-kafka-bridge: documentation/modules/proc-configuring-kafka-bridge-smr-metrics.adoc

> [id='proc-configuring-kafka-bridge-smr-metrics-{context}']

Reviewer: I don't know if the bridge has metrics dashboards but if so they probably need updating when using the metrics reporter.

[id='proc-configuring-kafka-bridge-smr-metrics-{context}']
= Configuring Strimzi Metrics Reporter metrics

[role="_abstract"]
Enable metrics for the Kafka Bridge by setting the `bridge.metrics` configuration.

.Prerequisites

* xref:proc-downloading-kafka-bridge-{context}[The Kafka Bridge installation archive is downloaded].

.Procedure

. Set the `bridge.metrics` configuration to `strimziMetricsReporter` and related configuration.
+
.Configuration for enabling metrics

[source,properties]
----
bridge.metrics=strimziMetricsReporter
----
+
Optionally, you can set a comma-separated list of regexes used to filter exposed metrics using the `kafka.prometheus.metrics.reporter.allowlist` property.
If not set, all available metrics will be exposed.
+

> You can add any plugin configuration to the Bridge properties file using the `kafka.` prefix.

Author: Yes, but we want metrics for all internal clients right?

> You can add any plugin configuration to the Bridge properties file using the `kafka.` prefix.

Reviewer 2: Yes, and is it something we want? Don't we want users able to have metrics for some clients but not others?

> You can add any plugin configuration to the Bridge properties file using the `kafka.` prefix.

Author: This is still possible using the client prefix to configure the `allowlist`. For example, to exclude all producer metrics the user can set `kafka.prometheus.metrics.reporter.allowlist=`.

> You can add any plugin configuration to the Bridge properties file using the `kafka.` prefix.

Reviewer 2: > to exclude all producer metrics the user can set kafka.prometheus.metrics.reporter.allowlist=.

I guess you meant `kafka.producer.prometheus.metrics.reporter.allowlist=.` ? Also should we say it clearly in the doc? Not just mentioning the `kafka.` prefix which covers all clients but also the other prefixes to cover specific clients?

[id='proc-configuring-kafka-bridge-smr-metrics-{context}']
= Configuring Strimzi Metrics Reporter metrics

[role="_abstract"]
Enable metrics with Strimzi Metrics Reporter by setting the `bridge.metrics` option to `strimziMetricsReporter`.

.Prerequisites

* xref:proc-downloading-kafka-bridge-{context}[The Kafka Bridge installation archive is downloaded].

.Procedure

. Set the `bridge.metrics` configuration to `strimziMetricsReporter`.
+
.Configuration for enabling metrics

[source,properties]
----
bridge.metrics=strimziMetricsReporter
----
+
Optionally, you can set a comma-separated list of regexes used to filter exposed metrics using the `kafka.prometheus.metrics.reporter.allowlist` property.

> If not set, all available metrics will be exposed.

Reviewer 3: Didn't we discussed in the past that we will have some more reasonable default than `.*`?

Author: That was discussed as part of the operator changes, but I think we can also apply that logic here.

Reviewer 3: So, we now have some different defaults in the code so this should be updated as well as I guess it is not all available metrics anymore?

> When needed, it is possible to configure the allowlist per client type.

Reviewer 4: We need some `+` adds to bring this content inline with step 1.

> When needed, it is possible to configure the allowlist per client type.

Reviewer 4: ```suggestion
When needed, it is possible to configure the `allowlist` per client type.
```

> For example, using the `kafka.admin` prefix, setting `kafka.admin.prometheus.metrics.reporter.allowlist=` excludes all admin client metrics.

Reviewer 4: ```suggestion
For example, by using the `kafka.admin` prefix and setting `kafka.admin.prometheus.metrics.reporter.allowlist=`, all admin client metrics are excluded.
```

> In case the same property is set with multiple prefixes, the most specific prefix wins.
> For example, `kafka.producer.prometheus.metrics.reporter.allowlist` wins over `kafka.prometheus.metrics.reporter.allowlist`. 

Reviewer 4: ```suggestion
+
In the event that the same property is set with multiple prefixes, the most specific prefix takes precedence.
For example, `kafka.producer.prometheus.metrics.reporter.allowlist` takes precedence over `kafka.prometheus.metrics.reporter.allowlist`. 
```

> You can add any plugin configuration to the Bridge properties file using `kafka.`, `kafka.admin.`, `kafka.producer.`, and `kafka.consumer.` prefixes.

Reviewer 4: ```suggestion
You can add any plugin configuration to the Kafka Bridge properties file using `kafka.`, `kafka.admin.`, `kafka.producer.`, and `kafka.consumer.` prefixes.
```

> You can add any plugin configuration to the Bridge properties file using `kafka.`, `kafka.admin.`, `kafka.producer.`, and `kafka.consumer.` prefixes.

Reviewer 4: should we move this above "If required, you can configure an `allowlist` for a specific client type."  as we use `kafka.admin' in the example there?

> 
> You can add any plugin configuration to the Bridge properties file using `kafka.`, `kafka.admin.`, `kafka.producer.`, and `kafka.consumer.` prefixes.

Reviewer 4: ```suggestion
+
You can add any plugin configuration to the Bridge properties file using `kafka.`, `kafka.admin.`, `kafka.producer.`, and `kafka.consumer.` prefixes.
```

> See the https://github.com/strimzi/metrics-reporter[Strimzi Metrics Reporter] documentation for more details.

Reviewer 3: What exactly does this mean, given we have only one set of metrics on the output?

Reviewer 2: The Strimzi metrics reporter plugin configuration is a per-client configuration.
The bridge has 3 clients, admin, producer and consumer and you can apply a configuration to all of them by using the `kafka.` prefix or a specific config for each client using the corresponding prefix (i.e. if you want a specific config on the consumer only, you set it with the `kafka.consumer.` prefix).
For example in the above description, you could have an allowList different for admin, producer and consumer by setting it with the appropriate prefix, for example:

````
kafka.consumer.prometheus.metrics.reporter.allowlist=<some list>"
kafka.producer.prometheus.metrics.reporter.allowlist=<some other list>"
```

Author: For example, I could only be interested in admin client metrics for troubleshooting purpose, so I would set producer and consumer to empty allow list.

Reviewer 3: Hmm, maybe this would deserve some clarification in the docs. Especially the `the most specific always wins` could be confusing as you can also take it to the list of the different expressions etc.

. Run the Kafka Bridge run script.
+
.Running the Kafka Bridge
[source,shell]
----
./bin/kafka_bridge_run.sh --config-file=<path>/application.properties
----
+

> With metrics enabled, you can scrape metrics in the Prometheus format from the `/metrics` endpoint of the Kafka Bridge.

Reviewer 4: ```suggestion
With metrics enabled, you can scrape metrics in Prometheus format from the `/metrics` endpoint of the Kafka Bridge.
```
